sql: code generation for missing pg catalog tables

mnovelodou opened the pull request:
Previously, programmer had to add missing table manually,
This was inadequate because postgress added a lot of tables and
manual process can lead to coding mistakes or gaps
To address this, diff tool now have the hability to generate the
missing code

Release note: None

Fixes #58001

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58519)
<!-- Reviewable:end -->

Commit:
sql: code generation for missing pg catalog tables

Previously, programmer had to add missing table manually,
This was inadequate because postgress added a lot of tables and
manual process can lead to coding mistakes or gaps
To address this, diff tool now have the hability to generate the
missing code

Release note: None

Fixes #58001

Release justification: non-production code changes

RichardJCai commented on pkg/sql/pg_catalog_diff.go:
```diff
@@ -185,6 +188,30 @@ func (p PGCatalogTables) rewriteDiffs(diffFile string) error {
 	return nil
 }
 
+// getNotImplementedTables retrieves tables that are not yet part of crdb
```
Rename to getUnimplementedTables

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/pg_catalog_diff.go:
```diff
@@ -185,6 +188,30 @@ func (p PGCatalogTables) rewriteDiffs(diffFile string) error {
 
+// getNotImplementedTables retrieves tables that are not yet part of crdb
+func (p PGCatalogTables) getNotImplementedTables(source PGCatalogTables) PGCatalogTables {
+	notImplemented := make(PGCatalogTables)
```
unimplemented

mnovelodou replied:
Done

RichardJCai commented on pkg/sql/pg_catalog_diff.go:
```diff
@@ -185,6 +188,30 @@ func (p PGCatalogTables) rewriteDiffs(diffFile string) error {
 	return nil
 }
 
+// getNotImplementedTables retrieves tables that are not yet part of crdb
```
Also in general please remember to add a period to comments (for correct grammar)

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/pg_catalog_diff.go:
```diff
@@ -185,6 +188,30 @@ func (p PGCatalogTables) rewriteDiffs(diffFile string) error {
 }
 
+// getNotImplementedTables retrieves tables that are not yet part of crdb
+func (p PGCatalogTables) getNotImplementedTables(source PGCatalogTables) PGCatalogTables {
```
Can you add a note stating where source is coming from so in the future we can easily tell what's being compared (crdb current tables, and pg as source)

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/pg_catalog_diff.go:
```diff
@@ -185,6 +188,30 @@ func (p PGCatalogTables) rewriteDiffs(diffFile string) error {
+}
+
+//AreAllTypesImplemented verifies that all the types are implemented in cockroach db
+func (c PGCatalogColumns) getNotImplementedTypes() map[oid.Oid]string {
```
unimplemented

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/pg_catalog_test.go:
```diff
@@ -28,32 +28,77 @@
+	allTableNamesClose       = `),`
+	virtualTablePosition     = `// typOid is the only OID generation approach that does not use oidHasher, because`
+	virtualTableTempl        = `var %s = virtualSchemaTable{
+	comment: "%s was created for compatibility",
```
Update comment to 
%s was created for compatibility and is currently unimplemented

(That's what I used when adding the missing tables)

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/pg_catalog_test.go:
```diff
@@ -170,9 +215,396 @@ func rewriteDiffs(t *testing.T, diffs PGCatalogTables, diffsFile string) {
 	}
 }
 
+// fixConstants helps to update catconstants that are needed for pgCatalog
```
fixConstants updates catconstants that are needed for pgCatalog

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/pg_catalog_test.go:
```diff
@@ -170,9 +215,396 @@ func rewriteDiffs(t *testing.T, diffs PGCatalogTables, diffsFile string) {
+		text := reader.Text()
+		trimText := strings.TrimSpace(text)
+
+		if strings.HasPrefix(trimText, "//") {
```
Can you leave a comment or add an example about why we need this branch

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/pg_catalog_test.go:
```diff
@@ -170,9 +215,396 @@ func rewriteDiffs(t *testing.T, diffs PGCatalogTables, diffsFile string) {
+// appendString calls WriteString and panics on error
+func (o outputFile) appendString(s string) {
+	if _, err := o.f.WriteString(s); err != nil {
+		panic(fmt.Errorf("error while writing string: %s: %v", s, err))
```
I think we can pass `t *testing` here and use t.Fatalf instead for a clearer error message

RichardJCai replied:
And in general we can replace all the panics with this

mnovelodou replied:
I am going to push back here:
If you notice, the whole purpose of outputFile type is to wrap io.File (for writers) and have a helper function that avoids any linter complain from the error returned by "WriteString". Which is super improbable that it happen and it would not happen in normal situations.

So I strongly believe that the refactoring of this method to receive *testing.T and using it instead of panic, it is a big effort and the benefit is not really worthy.

RichardJCai commented on pkg/sql/pg_catalog_test.go:
```diff
@@ -170,9 +215,396 @@ func rewriteDiffs(t *testing.T, diffs PGCatalogTables, diffsFile string) {
+}
+
+// getTableDefinitionsText retrieves pgCatalog.tableDefs, then it merges the missing table in that object and formats
+// the replacement text
```
Weird commenting here

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/virtual_schema.go:
```diff
@@ -176,6 +177,7 @@ func (t virtualSchemaTable) initVirtualTableDesc(
 		tree.PersistencePermanent,
 	)
 	if err != nil {
+		log.Errorf(ctx, "initVirtualDesc problem: %v\n%s", err, t.schema)
```
Do we need this extra here? Shouldn't the return err handle it, we can also wrap the error instead.

mnovelodou replied:
The problem of not having these is that the error is eventually wrapped and manipulated later in the code. During my tests I end up having a bad create table that I was unable to find which definition until I did some debugging of the code.

I strongly recommend to leave that log, in a production environment I don't believe it gets printed.

RichardJCai commented on pkg/sql/pg_catalog_test.go:
```diff
@@ -28,32 +28,77 @@
+	allTableNamesDeclaration = `allTableNames: buildStringSet(`
+	allTableNamesClose       = `),`
+	virtualTablePosition     = `// typOid is the only OID generation approach that does not use oidHasher, because`
+	virtualTableTempl        = `var %s = virtualSchemaTable{
```
Replace with virtualTableTemplate for clarity

mnovelodou replied:
Done.

RichardJCai commented on pkg/sql/pg_catalog_test.go:
```diff
@@ -28,32 +28,77 @@
+	pgCatalogIDConstant      = "PgCatalogID"
+	tableIDSuffix            = "TableID"
+	tableDefsDeclaration     = `tableDefs: map[descpb.ID]virtualSchemaDef{`
+	tableDefsClosure         = `},`
```
Rename to tableDefCloseBracket (not sure Closure is the right word here)

mnovelodou replied:
I named it closure because I was thinking of the function "printsBeforeClosure" probably closure is not the right word,
What printsBeforeClosure does:
It will read and skips all the lines (Except comments) until find the "closure" line, then it will print the s text and the "closure" text.

Please let me know if you believe there is a better word for this.

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58519)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou, @rafiss, and @RichardJCai)

___
*[pkg/sql/pg_catalog_diff.go, line 191 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MTHE1wP98i-8kQNUaPc:-MTHE1wP98i-8kQNUaPd:b-gcvrnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/962b06c6d380c9ad00fa2472ad60aa3cd7d9401a/pkg/sql/pg_catalog_diff.go#L191)):*
> ```Go
> }
> 
> // getNotImplementedTables retrieves tables that are not yet part of crdb
> ```

Would like this comment to be more specific, I guess in this case it would be retrieves PGCatalog tables that are not yet part of CRDB.

Also in general comments should finished off with periods (valid sentences)
___
*[pkg/sql/pg_catalog_diff.go, line 202 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MTHEFMu18R1YiekPyEr:-MTHEFMu18R1YiekPyEs:b-aklytg) ([raw file](https://github.com/cockroachdb/cockroach/blob/962b06c6d380c9ad00fa2472ad60aa3cd7d9401a/pkg/sql/pg_catalog_diff.go#L202)):*
> ```Go
> }
> 
> //AreAllTypesImplemented verifies that all the types are implemented in cockroach db
> ```



Missing space from comment to first word.

Also the function name should be updated in the comment.
___
*[pkg/sql/pg_catalog_test.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MQP3YjJ8E6yfZqkzcGK-r5-72:-MVMlEEv0r5WrM9jfxIg:br4eyig) ([raw file](https://github.com/cockroachdb/cockroach/blob/22c90455043b82fb489eb9b7a431615a3e4aa8b1/pkg/sql/pg_catalog_test.go#L72)):*
> Closure

I suggest we just name it `tableDefCloseBracket` it's more clear in my opinion.
___
*[pkg/sql/pg_catalog_test.go, line 373 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MQP3YjJ8E6yfZqkzcGK-r5-373:-MVMlS-uCjsVuO8ubSFe:b-9qtypw) ([raw file](https://github.com/cockroachdb/cockroach/blob/22c90455043b82fb489eb9b7a431615a3e4aa8b1/pkg/sql/pg_catalog_test.go#L373)):*
> push back
Sure, I'm okay with leaving it if it is a big effort.
___
*[pkg/sql/virtual_schema.go, line 180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MQP3YjKFIgiYh9AWtZX-r5-180:-MVMkgAsEugNPmWbXEBQ:b-fidr6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/22c90455043b82fb489eb9b7a431615a3e4aa8b1/pkg/sql/virtual_schema.go#L180)):*
<details><summary><i>Previously, mnovelodou (Miguel Novelo) wrote…</i></summary><blockquote>

The problem of not having this is that the error is eventually wrapped and manipulated later in the code. During my tests I end up having a bad create table that I was unable to find which definition until I did some debugging of the code.

I strongly recommend to leave that log, in a production environment I don't believe it gets printed.
</blockquote></details>

Instead of logging here, would it be alright to wrap the returned error?
We can do `errors.Wrapf(err, "initVirtualDesc problem: %s", t.schema)` 

If the returned error is unclear, I think we should make it clear instead of adding a log.


<!-- Sent from Reviewable.io -->

mnovelodou reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58519)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @RichardJCai)

___
*[pkg/sql/pg_catalog_diff.go, line 191 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MTHE1wP98i-8kQNUaPc:-MVNAg5tBBS9h3GIXQDC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/962b06c6d380c9ad00fa2472ad60aa3cd7d9401a/pkg/sql/pg_catalog_diff.go#L191)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Would like this comment to be more specific, I guess in this case it would be retrieves PGCatalog tables that are not yet part of CRDB.

Also in general comments should finished off with periods (valid sentences)
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog_diff.go, line 202 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MTHEFMu18R1YiekPyEr:-MVNB_BO6ArYFHvaLkKM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/962b06c6d380c9ad00fa2472ad60aa3cd7d9401a/pkg/sql/pg_catalog_diff.go#L202)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>



Missing space from comment to first word.

Also the function name should be updated in the comment.
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog_test.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MQP3YjJ8E6yfZqkzcGK-r5-72:-MVNBjPVEei1_0Qtkgl4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/22c90455043b82fb489eb9b7a431615a3e4aa8b1/pkg/sql/pg_catalog_test.go#L72)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

> Closure

I suggest we just name it `tableDefCloseBracket` it's more clear in my opinion.
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog_test.go, line 373 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MQP3YjJ8E6yfZqkzcGK-r5-373:-MVNIEVW7UahdrGmalTZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/22c90455043b82fb489eb9b7a431615a3e4aa8b1/pkg/sql/pg_catalog_test.go#L373)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

> push back
Sure, I'm okay with leaving it if it is a big effort.
</blockquote></details>

Done.
___
*[pkg/sql/virtual_schema.go, line 180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MQP3YjKFIgiYh9AWtZX-r5-180:-MVNICl9DEJErC22dq20:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/22c90455043b82fb489eb9b7a431615a3e4aa8b1/pkg/sql/virtual_schema.go#L180)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Instead of logging here, would it be alright to wrap the returned error?
We can do `errors.Wrapf(err, "initVirtualDesc problem: %s", t.schema)` 

If the returned error is unclear, I think we should make it clear instead of adding a log.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58519)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mnovelodou, @rafiss, and @RichardJCai)

___
*[pkg/sql/pg_metadata_test.go, line 78 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNJqkmCIZrTsL_MDuW:-MVNJqkmCIZrTsL_MDuX:bqeacjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/pg_metadata_test.go#L78)):*
> ```Go
> 	tableIDSuffix            = "TableID"
> 	tableDefsDeclaration     = `tableDefs: map[descpb.ID]virtualSchemaDef{`
> 	tableDefsCloseBracket    = `},`
> ```

Maybe we should name this tableDefsTerminal 
___
*[pkg/sql/pg_metadata_test.go, line 80 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNK0zaA_7KBbgak4sj:-MVNK0zaA_7KBbgak4sk:bvhzphk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/pg_metadata_test.go#L80)):*
> ```Go
> 	tableDefsCloseBracket    = `},`
> 	allTableNamesDeclaration = `allTableNames: buildStringSet(`
> 	allTableNamesClose       = `),`
> ```

allTablesNameTerminal
___
*[pkg/sql/pg_metadata_test.go, line 318 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNK5Pj3ij5LBDSdnpw:-MVNK5Pj3ij5LBDSdnpx:b3v3hpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/pg_metadata_test.go#L318)):*
> ```Go
> 
> // printBeforeTerminalString will skip all the lines and print `s` text when finds the terminal string.
> func printBeforeTerminalString(
> ```

I realized both "TerminalStrings" are actually chars (just a close bracket) I think we can rename this to `printBeforeTerminalChar`
___
*[pkg/sql/pg_metadata_test.go, line 675 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNJbzAE5kEf6LrF4lx:-MVNJbzB21FJFBEfGMwC:b65f35y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/pg_metadata_test.go#L675)):*
> ```Go
> 
> 	if *addMissingTables {
> 		notImplemented := diffs.getUnimplementedTables(pgTables)
> ```

notImplemented -> unimplemented
___
*[pkg/sql/virtual_schema.go, line 180 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNJWKvDfsoMWNRsabI:-MVNJWKw26r7fbz0ZBWY:b-5ahu6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/virtual_schema.go#L180)):*
> ```Go
> 	)
> 	if err != nil {
> 		err = errors.Wrapf(err, "initVirtualDesc problem: %v\n%s", err, t.schema)
> ```

I think this is wrapping the error twice. errors.Wrapf will wrap the error for you, so we don't need to pass the err the second time.


<!-- Sent from Reviewable.io -->

mnovelodou reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58519)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @RichardJCai)

___
*[pkg/sql/pg_metadata_test.go, line 78 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNJqkmCIZrTsL_MDuW:-MVSMZFh8rvoWJLZhsyR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/pg_metadata_test.go#L78)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Maybe we should name this tableDefsTerminal 
</blockquote></details>

Done.
___
*[pkg/sql/pg_metadata_test.go, line 80 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNK0zaA_7KBbgak4sj:-MVSO3oW3Gq-opt-9nt0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/pg_metadata_test.go#L80)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

allTablesNameTerminal
</blockquote></details>

Done.
___
*[pkg/sql/pg_metadata_test.go, line 318 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNK5Pj3ij5LBDSdnpw:-MVSMkW14FuJTVk3K1OQ:b3i7a2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/pg_metadata_test.go#L318)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I realized both "TerminalStrings" are actually chars (just a close bracket) I think we can rename this to `printBeforeTerminalChar`
</blockquote></details>

Not Really, tableDefsTerminal is `),` and all tableDefsTerminal is `),` (Both have commas) and they check the trimmed entire line (This is not looking at a single character).

If you look at the function this terminalString is compared with trimText which is a line read from the file without blank spaces at beginning and end of the string.
___
*[pkg/sql/pg_metadata_test.go, line 675 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNJbzAE5kEf6LrF4lx:-MVSNdn33GBJb9lmN1Vs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/pg_metadata_test.go#L675)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

notImplemented -> unimplemented
</blockquote></details>

Done.
___
*[pkg/sql/virtual_schema.go, line 180 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-MVNJWKvDfsoMWNRsabI:-MVSOYtb9lWXI2i_bmYv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d4ad8cee82b31e8014f02cf771a3248059b22b/pkg/sql/virtual_schema.go#L180)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I think this is wrapping the error twice. errors.Wrapf will wrap the error for you, so we don't need to pass the err the second time.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (approved):
One thing I just realized is that the `allTableNames` list for schema definitions is not actually the source of truth for what tables we implement. Ie we don't implement pg_sequence but it's listed there. 

I think this is fine to merge as is since tables that appear in `allTableNames` will return an unimplemented error if they're used and we already merged tables using this tool.

https://github.com/cockroachdb/cockroach/issues/61801 I opened an issue here.

What we can do to actually check which tables we implement is getting the tables using `NewVirtualSchemaHolder` in virtual_schema.go. This way we can check what actually works. 

Either way I'm okay with merging this and iterating since this solution implements the tables that we did not list in `allTableNames`, we just have to change the soltuion to also add tables that we don't implement but also appear in `allTableNames`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58519#-:-MVT-mOk-yU_9n568ZMK:bvlziz7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @RichardJCai)



<!-- Sent from Reviewable.io -->

RichardJCai commented:
bors r+
